### ISSUE TITLE
508-defect-1 ❗ [COGNITION, SEMANTIC MARKUP]: Form field instructions MUST be associated with inputs or buttons

### ISSUE THREAD

**Issue by joshkimux** [department-of-veterans-affairs/va.gov-team#19570]
# [508-defect-1 :exclamation: Launchblocker](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/accessibility/guidance/defect-severity-rubric.md#508-defect-1)

## Feedback framework

- **❗️ Must** for if the feedback must be applied 
- **⚠️ Should** if the feedback is best practice 
- **✔️ Consider** for suggestions/enhancements 

## Definition of done

1. Review and acknowledge feedback.
1. Fix and/or document decisions made.
1. Accessibility specialist will close ticket after reviewing documented decisions / validating fix.

## Point of Contact

**VFS Point of Contact:** _Josh_

## User Story or Problem Statement

As a screen reader user, I want to know which fields are grouped together so that I can have an equivalent experience to sighted users when navigating my direct deposit options.

## Details

> Using the right HTML elements when implementing forms is essential to ensure they can be used by as many people as possible including screen reader users.

Sighted users will be able to _visually_ determine that fields are grouped together thanks to the grey border. However, screen reader users will not have access to the same information for the following reasons:
- Upon clicking "edit," screen reader users may likely tab to the first item expecting a form field. Screen reader users will hear "Where can I find these numbers," which has no meaning on its own without a group label (What numbers?). The first set of instructions will be skipped, and will not be described to the user upon landing on the help accordion.
- Upon tabbing again and entering the first input field, users will hear "Routing number (Your bank’s name will appear after you add the 9-digit routing number)," but no other information to help contextualize the task. Screen reader users with cognitive disabilities may struggle to remember why they are providing this information.
- Upon tabbing again, users will hear "Account number (This should be no more than 17 digits)(*Required)." Screen reader users may struggle to understand that this is grouped with the first input field as they would not have heard the initial instructions "Please enter your bank’s routing and account numbers and your account type" while tabbing.
- Both update and cancel buttons do not have a unique accessible label as there are update and cancel buttons in the latter half the page.

For these collective reasons and more, I've chosen to write this as a launchblocking ticket. 

## Acceptance Criteria

- [ ] Fields are semantically grouped and labelled
- [ ] Buttons have unique accessible labels 

## Environment

* Operating System: any
* Browser: all
* Screenreading device: any
* Server destination: all

## Steps to Recreate

1. Enter `https://staging.va.gov/profile` in browser
2. Start any screenreader 
3. Tab to the first edit button and select it
4. Confirm the screen reader announces some variation of "Where can I find these numbers, button"-- but _does not_ announce a group label (e.g. edit bank account for disability compensation and pension benefits)

## Ideal Solution

Ideally, a preferred semantic approach to solving this problem would be to, upon activating edit mode,
- extend the `form` to before the `dfn` "Account" in the DOM
- wrap the `form` contents within a `fieldset`
- convert the `dfn` "Account" to a `legend` labelled "Edit bank account for disability compensation and pension benefits"
- ensure the `legend` is the first child of the `fieldset`

```html
<form>
  <fieldset>
    <legend> <!-- Must be the first child in order to be announced -->
      Edit bank account for disability compensation and pension benefits
    </legend
    <!-- fields go here -->
  </fieldset>
</form>
```

This would add a semantic group label to all fields within the fieldset. If a screen reader were to tab to the first button "Where can I find these numbers..." the screen reader would also announce, "edit bank account for disability compensation..." which contextualizes its use (bonus points if we can rewrite "where can I find these numbers" to "where can I find my bank account numbers" to meet link purpose). This would apply to all other fields, including the update and cancel buttons. We would not need to rely on `aria`, which meets the first rule of `aria` which is to _not_ use `aria` when it can be avoided with pure semantic html.

## Band Aid Solution

That being said, the ideal solution may require extensive changes to the form design system. An alternative, and less change-heavy method of meeting the acceptance criteria _might_ be to use `role="group"` and `aria-label`, but more testing would be needed.

```html
<div role="group" aria-label="Edit bank account for disability compensation and pension benefits">
  <span>
    <div id="errors"></div>
    <p class="vads-u-margin-top--0">Please enter your bank’s routing and account numbers and your account type.</p>
    <!-- etc -->
  </span>
</div>
```

## WCAG or Vendor Guidance (optional)

* [Using the fieldset and legend elements, gov.uk](https://accessibility.blog.gov.uk/2016/07/22/using-the-fieldset-and-legend-elements/)
* [Grouping controls, WCAG guidance](https://www.w3.org/WAI/tutorials/forms/grouping/)

## Screenshots or Trace Logs

<img width="1792" alt="Screen Shot 2021-02-05 at 1 51 08 PM" src="https://user-images.githubusercontent.com/14154792/107076821-f8759c00-67b9-11eb-9a66-300a887b53cc.png">
<img width="1792" alt="Screen Shot 2021-02-05 at 1 41 31 PM" src="https://user-images.githubusercontent.com/14154792/107076824-fad7f600-67b9-11eb-817d-9be18f921b65.png">

### Proposed Solution 
<img width="1717" alt="Screen Shot 2021-02-05 at 2 37 09 PM" src="https://user-images.githubusercontent.com/14154792/107080852-d0893700-67bf-11eb-86a0-3147153221d4.png">

https://user-images.githubusercontent.com/14154792/107080854-d252fa80-67bf-11eb-831a-083a019d6dc2.mov

**Comment by mshea0606** [comment]
@erikphansen update aria labels in buttons

**Comment by mshea0606** [comment]
@sporeboy @erikphansen it would state "Update your direct deposit information" OR "Update your education benefit direct deposit information". Confirm with @joshkimux

**Comment by mshea0606** [comment]
@erikphansen  this is listed as a must/launch blocker in Peggy's feedback

**Comment by erikphansen** [comment]
@joshkimux ready to validate on staging.

**Comment by mshea0606** [comment]
@1Copenut could you validate this since @joshkimux is out? Thanks!

**Comment by Samara-Strauss** [comment]
@joshkimux can we get a review on this ASAP so we can either make any required updates or close it out?

**Comment by joshkimux** [comment]
@Samara-Strauss On it now!

**Comment by Samara-Strauss** [comment]
Thank you!

**Comment by joshkimux** [comment]
<img width="752" alt="Screen Shot 2021-03-10 at 11 12 28 AM" src="https://user-images.githubusercontent.com/14154792/110660380-8239d000-8191-11eb-99b1-ac16a1a3af27.png">

Validated!
